### PR TITLE
feat: support Technicolor Vodafone Stations from Germany

### PR DESCRIPTION
**Pull request by guerda**
Add implementation for login, logout and get devices
Make `VodafoneStationCommonApi`  abstract and move some functions up and down
Make `library_test` support the different classes

### CODE REVIEW

File: src/aiovodafone/api.py [hunk 1]
```diff
@@ -118,10 +120,176 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+            bytes(self.password, "utf-8"),
+            bytes(salt_response["salt"], "utf-8"),
+            1000,
+        ).hex()[:32]
+        b = hashlib.pbkdf2_hmac(
+            "sha256",
+            bytes(a, "utf-8"),
+            bytes(salt_response["saltwebui"], "utf-8"),
+            1000,
+        ).hex()[:32]
```
Comment (chemelli74) [inline src/aiovodafone/api.py:191]: Can we plase move this to  "_encrypt_string()" like for Serecomm ?

**Comment by guerda** [comment]
Will expand #34 to implement this

**Review by chemelli74 — COMMENTED** [review]
First review run

File: src/aiovodafone/api.py [hunk 2]
```diff
@@ -118,10 +120,176 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
         _LOGGER.debug("GET reply %s: %s", page, reply_json)
         return await self._list_2_dict(reply_json)
 
+    async def convert_uptime(self, uptime: str) -> datetime:
+        """Convert router uptime to last boot datetime."""
+        d = int(uptime.split(":")[0])
+        h = int(uptime.split(":")[1])
+        m = int(uptime.split(":")[2])
+
+        return datetime.utcnow() - timedelta(days=d, hours=h, minutes=m)
```
[the remark this answers is not in the document]
Comment (guerda) [reply src/aiovodafone/api.py:129]: It is used in the library_test.py. I'll convert it into an abstract method then.

[the remark this answers is not in the document]
Comment (guerda) [reply src/aiovodafone/api.py]: Done

File: src/aiovodafone/api.py [hunk 3]
```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+    async def logout(self) -> None:
+        pass
+
 
 class VodafoneStationTechnicolorApi(VodafoneStationCommonApi):
     """Queries Vodafone Station running Technicolor firmware."""
 
+    def __init__(self, host: str, username: str, password: str) -> None:
+        super().__init__(host, username, password)
+        self.headers = {
```
Comment (chemelli74) [inline src/aiovodafone/api.py:153]: Can we use a common headers between the 2 class ?

File: src/aiovodafone/api.py [hunk 4]
```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+
+        Args:
+            credential (str): login password for the user
+            salt (str): salt given by the login response
+            salt_web_ui (str): salt given by the web UI
+
+        Returns:
+            str: the hash for the session API
+        """
+        _LOGGER.debug("Calculate hash")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:174]:
```suggestion
        _LOGGER.debug("Calculate credential hash")
```

File: src/aiovodafone/api.py [hunk 5]
```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+            1000,
+        ).hex()[:32]
+        return b
+
+    async def login(self) -> bool:
+        """Router login."""
+        _LOGGER.debug("Logging into %s", self.host)
+        _LOGGER.debug("Get salt for login")
+        page = self.host + "/api/v1/session/login"
+        reply = await self.session.post(
```
Comment (chemelli74) [inline src/aiovodafone/api.py:194]: e should use the common `post` method

File: src/aiovodafone/api.py [hunk 6]
```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+        salt = salt_response["salt"]
+        salt_web_ui = salt_response["saltwebui"]
+
+        # Calculate hash
+        password_hash = await self._encrypt_string(self.password, salt, salt_web_ui)
+
+        # Perform login
+        _LOGGER.debug("Perform login")
+        page = self.host + "/api/v1/session/login"
+        reply = await self.session.post(
```
Comment (chemelli74) [inline src/aiovodafone/api.py:211]: Same as before

File: src/aiovodafone/api.py [hunk 7]
```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
...
+            if login_response["message"] == "MSG_LOGIN_150":
+                raise AlreadyLogged
+            _LOGGER.error(await reply.text())
+            raise CannotAuthenticate
+
+        # Request menu otherwise the next call fails
+        _LOGGER.debug("Get menu")
+        now = int(time.time() * 1000)
+        page = self.host + "/api/v1/session/menu?_=" + str(now)
+        reply = await self.session.get(
```
Comment (chemelli74) [inline src/aiovodafone/api.py:229]: We should use a common `get` method

**Review by chemelli74 — CHANGES_REQUESTED** [review]
Second run

File: src/aiovodafone/api.py (hunk 7 above)
Comment (guerda) [reply src/aiovodafone/api.py:229]: I thought about it, too. The common method adds a CSRF token which is not needed in this implementation. Not sure how we want to handle the different parameters. Basically we would end up in different implementations for both. I could see how I can extract a generalized method.

File: src/aiovodafone/api.py (hunk 5 above)
Comment (guerda) [reply src/aiovodafone/api.py:194]: I'll check that.

File: src/aiovodafone/api.py (hunk 7 above)
Comment (chemelli74) [reply src/aiovodafone/api.py:229]: We can always have a bool to specify if CSRF should be appended or not. But at least we don't repeat the same code all over the libray.

**Comment by chemelli74** [comment]
Base refactor merged with main; please rebase.

**Comment by guerda** [comment]
I had to refactor a lot, as the typing to "application/html" is not needed with the Technicolor devices.
That's why I added a param `use_html_content_type`  which defaults to `False`. 
Also I needed to add a `raw` param to `_get_page_result` and `convert_to_dict` to allow skipping the conversion. Technicolor doesn't need that.

Now ` VodafoneStationTechnicolorApi` uses all the common methods in the API.

I will rebase this tomorrow, have to figure out that with a clear head.

**Comment by guerda** [comment]
Replaced by #43

**Pull request closed without merging**